chat.llamas.cardlogic.game.api/src/main/java/chat/llamas/cardlogic/game/api/GameEventMediatorInterface.java

```diff
@@ -13,30 +13,125 @@
+import chat.llamas.cardlogic.domain.commands.CommandDelegate;
+import chat.llamas.cardlogic.domain.commands.FoldCommand;
 import chat.llamas.cardlogic.domain.commands.JoinGameCommand;
 import chat.llamas.cardlogic.domain.commands.LeaveGameCommand;
 import chat.llamas.cardlogic.domain.commands.MessageCommand;
 import chat.llamas.cardlogic.domain.commands.StartGameCommand;
+import chat.llamas.cardlogic.domain.game.PlayerInterface;
+import chat.llamas.cardlogic.domain.game.RoundType;
 
 public interface GameEventMediatorInterface {
 	GameInterface getGame();
 	
-	default void startGame(StartGameCommand command) throws UnsupportedOperationException {
+	default void startGame(StartGameCommand command) throws OperationNotSupportedException {
 		if (getGame().getGameState() == GameState.INACTIVE) {
 			//fireStartGame(command);
 		}
 		else {
-			throw new UnsupportedOperationException("Game state must be inactive.");
+			throw new OperationNotSupportedException("Game state must be inactive.");
 		}
 	}
 	
-	void sendMessage(MessageCommand command) throws UnsupportedOperationException;
+	void sendMessage(MessageCommand command) throws OperationNotSupportedException;
 	
-	void cancelGame(CancelGameCommand command) throws UnsupportedOperationException;
+	default void cancelGame(CancelGameCommand command) throws OperationNotSupportedException {
+		if (getGame().getGameState() == GameState.PENDING) {
+			if (getGame().getStartingPlayer().isPresent()) {
+				RoundInterface round = getGame().getCurrentRound();
+				
+				if (Objects.equals(round.getRoundType(), RoundType.PRE_GAME)) {
+					if (Objects.equals( getGame().getStartingPlayer().get().getName(), command.getDelegate().getPlayerName())) {
+						// TODO: Cancel the game
+					}
+					else {
+						throw new OperationNotSupportedException("Game can only be cancelled by the player who started it.");
+					}
+				}
+				else {
+					throw new OperationNotSupportedException("Game is already in progress.");
+				}
+			}
+			else {
+				throw new OperationNotSupportedException("There is no starting player.");
+			}
+		}
+		else {
+			throw new OperationNotSupportedException("Game state must be pending to cancel.");
+		}
+	}
+
+	default void leaveGame(LeaveGameCommand command) throws OperationNotSupportedException {
+		if (Objects.equals(getGame().getGameState(), GameState.PENDING)) {
+			// The game state is still pending,
+			// just remove the player.
+			removePlayerFromGame(command);
+			
+		}
+		else if (Objects.equals(getGame().getGameState(), GameState.ACTIVE)) {
+			// Fold the player if they haven't already,
+			// and then remove the player.
+			
+			try {
+				getGame().acceptCommand(
+					String.join(" ",
+						FoldCommand.COMMAND_NAME,
+						CommandDelegate.PLAYER_FLAG_LONG +
+						GameConstants.ARGUMENT_SEPARATOR +
+						command.getDelegate().getPlayerName()));
+				
+				removePlayerFromGame(command);
+			}
+			catch (OperationNotSupportedException e) {
+				throw e;
+			}
```

Comment:
Now that you mention it, I'm not sure. To avoid recursion, perhaps we should have a version of `Game.acceptCommand` that accepts an actual Command object, so that only an external consumer will ever call `Game.acceptCommand(String)`